### PR TITLE
FRW-1877 Fixed the Queue Worker stopping too early when using a "stop-when-empty" option.

### PR DESCRIPTION
**Pull request by aleksandr-velikanov**
- Developer(s): @aleksandr-velikanov

- Ticket: https://spryker.atlassian.net/browse/FRW-1877

- Release Group: https://release.spryker.com/release-groups/view/5484

- PR Overview: https://release.spryker.com/release/pull-request/11047

- merge: squash

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   RabbitMq                 | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Introduced `RabbitMqQueueMessageCheckerPlugin` to check if there are messages in queues for RabbitMQ.
- Introduced `RabbitMqFacade::areQueuesEmpty()` to go through queues and see if at least one has messages.

### CODE REVIEW

**Review by kraal-spryker — CHANGES_REQUESTED** [review]
.

File: src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+namespace Spryker\Zed\RabbitMq\Communication\Plugin;
+
+use Spryker\Client\RabbitMq\Model\RabbitMqAdapter;
+use Spryker\Zed\Kernel\Communication\AbstractPlugin;
+use Spryker\Zed\QueueExtension\Dependency\Plugin\QueueMessageCheckerPluginInterface;
+
+/**
+ * {@inheritDoc}
```
Comment (kraal-spryker) [inline src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php:15]: pls remove inheritdoc and api tags here

**Review by kraal-spryker — CHANGES_REQUESTED** [review]
.

**Review by kraal-spryker — APPROVED** [review]

File: src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace Spryker\Zed\RabbitMq\Communication\Plugin;
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php:8]: > Plugin namespace, should end with the folder with the same name as module that we want to extend. (e.g. Plugin\Product) They should't be put to the Plugin folder directly.

https://spryker.atlassian.net/wiki/spaces/Framework/pages/3580395659/Common+CR+issues

File: src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * @method \Spryker\Zed\RabbitMq\Business\RabbitMqFacadeInterface getFacade()
+ * @method \Spryker\Zed\RabbitMq\RabbitMqConfig getConfig()
+ */
+class RabbitMqQueueMessageCheckerPlugin extends AbstractPlugin implements QueueMessageCheckerPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     *
+     * - Checks if any of the queues provided has messages in it.
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php:23]: Please fix with Grammarly

File: src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Communication/Plugin/RabbitMqQueueMessageCheckerPlugin.php:6]: [This](https://github.com/spryker/rabbit-mq/blob/master/.license) license does not match what we have in https://github.com/spryker/rabbit-mq/blob/master/LICENSE

Please align with the business which license we should have for this module, and either update https://github.com/spryker/rabbit-mq/blob/master/.license (and all file headers - we have a console command for that) or https://github.com/spryker/rabbit-mq/blob/master/LICENSE

**Review by asmarovydlo — CHANGES_REQUESTED** [review]
1. Changelogs are not actual, please fix
2. PRs and RG names are not valid, please see https://spryker.atlassian.net/wiki/spaces/Framework/pages/3580395659/Common+CR+issues
> PR/RG names do follow conventions
[Core Release Process | CoreReleaseProcess PRInformation](https://spryker.atlassian.net/wiki/spaces/CORE/pages/99418680#CoreReleaseProcess-PRInformation)

CC @kraal-spryker

**Comment by asmarovydlo** [comment]
Please (add/extend existing) tests

File: src/Spryker/Zed/RabbitMq/Business/RabbitMqBusinessFactory.php [hunk 5]
```diff
@@ -36,7 +36,7 @@ public function createQueue()
     /**
      * @return \Spryker\Zed\RabbitMq\Business\Model\Queue\QueueInfoInterface
      */
-    protected function createQueueInfo()
+    public function createQueueInfo()
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Business/RabbitMqBusinessFactory.php:39]:
```suggestion
    public function createQueueInfo(): QueueInfoInterface
```

File: src/Spryker/Zed/RabbitMq/Business/RabbitMqFacade.php [hunk 6]
```diff
@@ -82,4 +82,18 @@ public function setUserPermissions(LoggerInterface $logger)
...
+    /**
+     * {@inheritDoc}
+     *
+     * @api
+     *
+     * @param array<string> $queueNames
+     *
+     * @return bool
+     */
+    public function areQueuesEmpty($queueNames): bool
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Business/RabbitMqFacade.php:95]:
```suggestion
    public function areQueuesEmpty(array $queueNames): bool
```

File: src/Spryker/Zed/RabbitMq/Business/RabbitMqFacadeInterface.php [hunk 7]
```diff
@@ -71,4 +71,16 @@ public function setupConnection(): void;
...
+     * Specification:
+     *  - Checks if any of the applicable queues have messages.
+     *
+     * @api
+     *
+     * @param array<string> $queueNames
+     *
+     * @return bool
+     */
+    public function areQueuesEmpty($queueNames): bool;
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Business/RabbitMqFacadeInterface.php:85]:
```suggestion
    public function areQueuesEmpty(array $queueNames): bool;
```

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @return void
+     */
+    protected function _before()
+    {
+        $this->rabbitMqBusinessFactoryMock = $this->createMock(RabbitMqBusinessFactory::class);
+        $this->queueInfoMock = $this->createMock(QueueInfo::class);
+
+        $this->rabbitMqFacade = new RabbitMqFacade();
+        $this->rabbitMqFacade->setFactory($this->rabbitMqBusinessFactoryMock);
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:52]: Please create them where they are used, since object are changed by reference this approach leads to errors (when several tests modify those objects) that are hard to catch.

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    /**
+     * @return void
+     */
+    public function testAreQueuesEmpty()
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:58]: 1. Missing return type.
2. Please fix test names according to https://spryker.atlassian.net/wiki/spaces/Framework/pages/3580395659/Common+CR+issues

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        $this->queueInfoMock
+            ->method('areQueuesEmpty')
+            ->with($queueNames)
+            ->willReturn($expectedResult);
+
+        $result = $this->rabbitMqFacade->areQueuesEmpty($queueNames);
+
+        $this->assertEquals($expectedResult, $result);
+    }
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:75]: Please add // Act // Arrange // Assert annotations, please see examples in the existing modules.

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php (hunk 9 above)
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:58]: Please add tests for negative case

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            ->willReturn($this->queueInfoMock);
+
+        $this->queueInfoMock
+            ->method('areQueuesEmpty')
+            ->with($queueNames)
+            ->willReturn($expectedResult);
+
+        $result = $this->rabbitMqFacade->areQueuesEmpty($queueNames);
+
+        $this->assertEquals($expectedResult, $result);
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:74]: This test does not test anything, please create real queues and check on them.

File: src/Spryker/Zed/RabbitMq/Business/Model/Queue/QueueInfo.php [hunk 12]
```diff
@@ -81,4 +105,15 @@ protected function addRabbitMqQueues(RabbitMqQueueCollectionTransfer $rabbitMqQu
...
         return $rabbitMqQueueCollectionTransfer;
     }
+
+    /**
+     * @param string $currentQueueName
+     * @param array<string> $queueNames
+     *
+     * @return bool
+     */
+    public function isApplicableQueue(string $currentQueueName, array $queueNames): bool
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Business/Model/Queue/QueueInfo.php:115]: 1. This method is used only in this class, please make it protected and remove it from interface.
2. Since the method name is not very informative,  pleas get rid of it and use in_array instead.

File: src/Spryker/Zed/RabbitMq/Communication/Plugin/Queue/RabbitMqQueueMessageCheckerPlugin.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+     * @return bool
+     */
+    public function areQueuesEmpty(array $queueNames): bool
+    {
+        return $this->getFacade()->areQueuesEmpty($queueNames);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Communication/Plugin/Queue/RabbitMqQueueMessageCheckerPlugin.php:38]: [Minor]

```suggestion
```

File: src/Spryker/Zed/RabbitMq/Communication/Plugin/Queue/RabbitMqQueueMessageCheckerPlugin.php [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * @method \Spryker\Zed\RabbitMq\Business\RabbitMqFacadeInterface getFacade()
+ * @method \Spryker\Zed\RabbitMq\RabbitMqConfig getConfig()
+ */
+class RabbitMqQueueMessageCheckerPlugin extends AbstractPlugin implements QueueMessageCheckerPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     *
```
Comment (asmarovydlo) [inline src/Spryker/Zed/RabbitMq/Communication/Plugin/Queue/RabbitMqQueueMessageCheckerPlugin.php:22]: [Minor]
```suggestion
```

**Review by asmarovydlo — CHANGES_REQUESTED** [review]

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+        for ($i = 0; $i < static::MESSAGE_AMOUNT; $i++) {
+            $messagesToSend[] = $this->tester->buildSendMessageTransfer();
+        }
+
+        foreach (static::QUEUE_NAMES as $queueName) {
+            $queueClient->sendMessages($queueName, $messagesToSend);
+        }
+
+        sleep(10);
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:112]: we don't have asynchronous operations there, if we need it, than smth went wrong.
We shouldn't use this approach because it increased the amount of flickery tests.
Please get rid if it here & elsewhere in the PR.

**Review by asmarovydlo — CHANGES_REQUESTED** [review]

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php (hunk 15 above)
Comment (aleksandr-velikanov) [reply tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php]: the timeout left from the previous iteration, not necessary now, I'll remove it

**Review by asmarovydlo — CHANGES_REQUESTED** [review]

File: src/Spryker/Zed/RabbitMq/RabbitMqDependencyProvider.php [hunk 16]
```diff
@@ -61,6 +67,20 @@ protected function addQueueAdapter(Container $container)
...
+     */
+    protected function addQueueClient(Container $container)
+    {
+        $container->set(static::CLIENT_QUEUE, function () use ($container) {
+            return $container->getLocator()->queue()->client();
+        });
+
+        return $container;
+    }
+
```
[the remark this answers is not in the document]
Comment (aleksandr-velikanov) [reply src/Spryker/Zed/RabbitMq/RabbitMqDependencyProvider.php:83]: It is also used in the model: vendor/spryker/rabbit-mq/src/Spryker/Zed/RabbitMq/Business/Model/Queue/QueueInfo.php

File: src/Spryker/Zed/RabbitMq/Business/RabbitMqBusinessFactory.php [hunk 17]
```diff
@@ -36,13 +37,14 @@ public function createQueue()
...
      */
-    protected function createQueueInfo()
+    public function createQueueInfo(): QueueInfoInterface
     {
         return new QueueInfo(
             $this->getGuzzleClient(),
             $this->getConfig()->getApiQueuesUrl(),
             $this->getConfig()->getApiUsername(),
             $this->getConfig()->getApiPassword(),
+            $this->getQueueClient(),
```
Comment (annakotiuk) [inline src/Spryker/Zed/RabbitMq/Business/RabbitMqBusinessFactory.php:47]: Use adapter

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * @var \PHPUnit\Framework\MockObject\MockObject|\Spryker\Zed\RabbitMq\Business\RabbitMqBusinessFactory
+     */
+    protected $rabbitMqBusinessFactoryMock;
+
+    /**
+     * @var \PHPUnit\Framework\MockObject\MockObject|\Spryker\Zed\RabbitMq\Business\Queue\QueueInfo
+     */
+    protected $queueInfoMock;
+    /**
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:42]: please fix indentation

File: tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+        $this->rabbitMqFacade = new RabbitMqFacade();
+        $this->rabbitMqFacade->setFactory($this->rabbitMqBusinessFactoryMock);
+    }
+
+    /**
+     * @return void
+     */
+    public function testAreQueuesEmptyReturnsTrueForEmptyQueues(): void
+    {
+        $expectedResult = true;
```
Comment (asmarovydlo) [inline tests/SprykerTest/Zed/RabbitMq/Business/RabbitMqFacadeTest.php:64]: please add arrange/act/assert annotations

**Review by asmarovydlo — APPROVED** [review]
